### ISSUE TITLE
Cast of a variable to an IRI does not match in BGP but works with filter

### ISSUE THREAD

**Issue by JervenBolleman** [openlink/virtuoso-opensource#731]
```sparql
PREFIX up:<http://purl.uniprot.org/core/> 
SELECT ?start ?protein ?input
WHERE
{
	BIND('P05067' AS ?input)
	BIND(
      IF(strstarts(?input, 'PRO_'),
      	IRI(CONCAT('http://purl.uniprot.org/annotation/', ?input)), 
        IF (contains(?input, '-'), 
            IRI(CONCAT('http://purl.uniprot.org/isoform/', ?input)),
            IRI(CONCAT('http://purl.uniprot.org/uniprot/', ?input))
        )
      ) AS ?start)
  {
    BIND(?start AS ?protein)
    ?protein a up:Protein .
  #  FILTER(sameTerm(?protein, ?start))
  }
} OFFSET 0 LIMIT 1
```
Fails to return results, while the equivalent where the `BIND(?start AS ?protein)` is commented out and the `FILTER(sameTerm(?protein, ?start))` commented in does.

On <http://lod.openlinksw.com/sparql/>, both seem to fail the "Strict checking of void variables" check.

**Comment by HughWilliams** [comment]
If I move the `BIND` call below the triple pattern (i.e., where the `FILTER` clause is), then it does return a row same as the `FILTER` clause does, i.e. —

```sparql
PREFIX up:<http://purl.uniprot.org/core/> 
SELECT ?start ?protein ?input
WHERE
{
	BIND('P05067' AS ?input)
	BIND(
      IF(strstarts(?input, 'PRO_'),
      	IRI(CONCAT('http://purl.uniprot.org/annotation/', ?input)), 
        IF (contains(?input, '-'), 
            IRI(CONCAT('http://purl.uniprot.org/isoform/', ?input)),
            IRI(CONCAT('http://purl.uniprot.org/uniprot/', ?input))
        )
      ) AS ?start)
  {
    ?protein a up:Protein .
    BIND(?start AS ?protein)
   # FILTER(sameTerm(?protein, ?start))
  }
} OFFSET 0 LIMIT 1
```

**Comment by JervenBolleman** [comment]
@HughWilliams yes, that returns a result but is not guaranteed to be correct. Also if you have a more stringent sparql parser one will get an error like  `BIND clause alias '{}' was previously used` or `BIND: Variable used when already in-scope: ?protein in BIND(?start AS ?protein)`. You used virtuoso specific syntax to override the value of `?protein` not to join on it. See e.g. the results [here](http://lod.openlinksw.com/sparql/?default-graph-uri=&query=PREFIX+up%3A%3Chttp%3A%2F%2Fpurl.uniprot.org%2Fcore%2F%3E+%0D%0ASELECT+%3Fstart+%3Fprotein+%3Finput%0D%0AWHERE%0D%0A%7B%0D%0A%09BIND%28%27P05067%27+AS+%3Finput%29%0D%0A%09BIND%28%0D%0A++++++IF%28strstarts%28%3Finput%2C+%27PRO_%27%29%2C%0D%0A++++++%09IRI%28CONCAT%28%27http%3A%2F%2Fpurl.uniprot.org%2Fannotation%2F%27%2C+%3Finput%29%29%2C+%0D%0A++++++++IF+%28contains%28%3Finput%2C+%27-%27%29%2C+%0D%0A++++++++++++IRI%28CONCAT%28%27http%3A%2F%2Fpurl.uniprot.org%2Fisoform%2F%27%2C+%3Finput%29%29%2C%0D%0A++++++++++++IRI%28CONCAT%28%27http%3A%2F%2Fpurl.uniprot.org%2Funiprot%2F%27%2C+%3Finput%29%29%0D%0A++++++++%29%0D%0A++++++%29+AS+%3Fstart%29%0D%0A++%7B%0D%0A++++%3Fprotein+a+up%3AProtein+.%0D%0A++++BIND%28%3Fstart+AS+%3Fprotein%29%0D%0A+++%23+FILTER%28sameTerm%28%3Fprotein%2C+%3Fstart%29%29%0D%0A++%7D%0D%0A%7D+OFFSET+0+LIMIT+1&format=text%2Fhtml&timeout=30000&run=+Run+Query+)

**Comment by HughWilliams** [comment]
@JervenBolleman: OK, I shall have development look into and comment on this ...

**Comment by iv-an-ru** [comment]
Hello Jerven,

The SPARQL queries can be calculated from inner groups to outer, so first `BIND(?start AS ?protein)` uses an unbound variable `?start` making `?protein` unbound, then `?protein a up:Protein .` binds nothing because there is no such thing as triple without subject. The rest is not important --- even if `?start` is eventually bound, that can not increment the (zeroed) number of results.

With `FILTER` instead of `BIND`, the optimizer may eliminate the needless `{}` around `?protein a up:Protein .` BGP first, so you may get what you want.

The right policy is to avoid redundant `{}` and treat signals about "void" and "weird" variables as any other errors, even if they can be disabled.

Best Regards,
Ivan Mikhailov
OpenLink Software

**Comment by JervenBolleman** [comment]
Hi Ivan,

The basic query below fails the strict checking of void variables.

```sparql
PREFIX up:<http://purl.uniprot.org/core/>
SELECT ?protein ?input
WHERE {
    BIND('P05067' AS ?input) 
    BIND(IRI(CONCAT('http://purl.uniprot.org/uniprot/', ?input)) AS ?protein) 
    ?protein a up:Protein .
} OFFSET 0 LIMIT 1
```

Also the following fails to return a result
```
PREFIX up:<http://purl.uniprot.org/core/>
SELECT ?protein ?input
WHERE {
  OPTIONAL {    
    FILTER(!BOUND(?protein))
    BIND('P05067' AS ?input) 
    BIND(IRI(CONCAT('http://purl.uniprot.org/uniprot/', ?input)) AS ?protein) 
    ?protein a up:Protein .
  }
  FILTER(BOUND(?protein))
  ?protein a up:Protein .
} 
```
That looks almost correct but I think it is not. This optimization seems to assume the FILTER applies to the whole OPTIONAL group but it should not. The optional group is 3 BGP sets by my counting not one. This because a BIND closes an BGP. Otherwise yes, a FILTER applies to the whole optional group.

But you are correct the only valid query solving our user case is this
```sparql
PREFIX skos:<http://www.w3.org/2004/02/skos/core#> 
PREFIX up:<http://purl.uniprot.org/core/> 
SELECT 
    ?protein 
	?organism 
	(GROUP_CONCAT(?geneLabel;separator=" , ") AS ?geneLabels)
WHERE
{
 OPTIONAL {
    BIND('_INPUT_' AS ?input)
    FILTER(strstarts(?input, 'PRO_'))
    BIND(IRI(CONCAT('http://purl.uniprot.org/annotation/', ?input)) AS ?annotation)
    ?proteinA up:annotation ?annotation .
  }
  OPTIONAL {
    BIND('_INPUT_' AS ?input)
    FILTER(contains(?input, '-'))
    FILTER(!BOUND(?proteinA))
    BIND(IRI(CONCAT('http://purl.uniprot.org/isoforms/', ?input)) AS ?sequence)
    ?proteinS up:sequence ?sequence .
  } 
  OPTIONAL {
    BIND('_INPUT_' AS ?input)
    FILTER(!BOUND(?proteinA) && !BOUND(?proteinS))
    BIND(IRI(CONCAT('http://purl.uniprot.org/uniprot/', ?input)) AS ?proteinP)
    ?proteinP a up:Protein .
  }
  BIND(IF(BOUND(?proteinA), ?proteinA,IF(BOUND(?proteinS), ?proteinS, ?proteinP)) AS ?protein)
  ?protein up:organism ?organism ;
           up:encodedBy/skos:prefLabel ?geneLabel .
} 
GROUP BY ?protein ?organism
```
Where the `_INPUT_` is modified by an external program.
